Added AudioTrack, AudioTrackList, VideoTrack, and VideoTrackList inte…

sreeise opened the pull request:
…rfaces

<!-- Please describe your changes on the following line: -->

Added AudioTrack, AudioTrackList, VideoTrack, VideoTrackList, and TrackEvent interfaces to support multiple media tracks.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22301  (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22622)
<!-- Reviewable:end -->

sreeise committed:
Added AudioTrack, AudioTrackList, VideoTrack, VideoTrackList, and TrackEvent interfaces

sreeise commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -200,6 +202,10 @@ pub struct HTMLMediaElement {
     #[ignore_malloc_size_of = "Rc"]
     played: Rc<DomRefCell<TimeRangesContainer>>,
+    // https://html.spec.whatwg.org/multipage/#dom-media-audiotracks
+    audio_tracks_list: MutNullableDom<AudioTrackList>,
```
Not sure whether MutNullableDom is correct here or whether some other area is missing for tracks. This would cause a build failure for attempting to share JSObject between threads when removing the audio and video tracks from the tracks lists in the dedicated media source failure steps. Step 2:

https://html.spec.whatwg.org/multipage/media.html#dedicated-media-source-failure-steps

sreeise replied:
Ahh nevermind, I was referring to here:

https://github.com/servo/servo/blob/bffe2a699e9f0b2c41aa811ea2c3e90e40342419/components/script/dom/htmlmediaelement.rs#L788

but didnt notice the this.root() assignment for 'this' here:

https://github.com/servo/servo/blob/bffe2a699e9f0b2c41aa811ea2c3e90e40342419/components/script/dom/htmlmediaelement.rs#L775

jdm reviewed (changes requested):
Since the code for the audio and video tracks is so similar, please consider whether each comment made about the video track list implementation could be applied to the audio track list implementation.  My one concern about this PR is that we're adding a bunch of code that does not get executed, since there are no callers for AudioTrackList::add and VideoTrackList::add. You can get the information about audio and video tracks from the Metadata value that's in PlayerEvent::MetadataUpdated.

sreeise commented:
> Welp, I totally forgot to read the long useful description in the pull request that you provided, but I think we came to similar conclusions.

No worries! You actually answered all of questions I had and was actually going to ask now so it works for me. Thanks for the helpful information and ill get back with a much cleaner pr soon :).

sreeise commented:
@jdm 
Hey Josh, I am trying to figure out how to update the tracks correctly but unsure on this. When looking at the metadata from here:

https://github.com/servo/servo/blob/6197f1addad631455cb4dcb215e172896f54a9ab/components/script/dom/htmlmediaelement.rs#L1134

It seems that only the `metadata.format` would be relevant (besides the codecs you mentioned) from this part? I was trying to use GDB to figure out what exactly the metadata format can look like to determine the best route but could not debug it successfully. Is the format showing the type such as "audio" or "video" or is there something more specific here? 

Also, is the correct way to retrieve the rest of the track values by getting the attribute with the local_name! macro? Something like:

`element.get_string_attribute(&local_name!("kind"));`

Thanks for the help!

jdm commented:
The format contains a string that corresponds with https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gst-plugins-bad-libs/html/GstPlayerMediaInfo.html#gst-player-media-info-get-container-format, which doesn't give any meaningful information about what it might contain. I'm not sure that it will be useful here.

If the media tracks have a `kind` content attribute, then that would be the appropriate way to retrieve if it `element` is an HTMLTrackElement (or an Element that is actually an HTMLTrackElement under the hood).

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22692) made this pull request unmergeable. Please resolve the merge conflicts.

sreeise commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1233,74 @@ impl HTMLMediaElement {
             },
             PlayerEvent::MetadataUpdated(ref metadata) => {
```
On the Kind attribute:
On determining the kind it seems there may not be enough information but I could very well be wrong here as well.

Metadata.format is showing the audio/video type such as 'ogg' where as metadata.audio_tracks / video_tracks shows the specific type such as 'vorbis' for ogg audio. When trying to add the kind to a track it is more specific such as audio/alternate or audio/audiodesc for ogg.

If I am reading this correctly, in the specs that are referenced for the ogg format: [skeleton headers]( https://wiki.xiph.org/SkeletonHeaders),  this would work as long as the tracks are in the correct order. In this case, the first Vorbis audio would be the main audio or main-desc. However, there is also a Vorbis audio for audio descriptions. I wonder if there is only one type of audio for Vorbis would it just be assumed to be the main audio.

DASH seems to be the same way. The spec specifies a role scheme in the form of: urn:mpeg dash:role 2011
This would also require knowing the role type as well in the same metadata.

AC3 and MPEG-2 require bsmod and full_svc values to determine the the correct kind. I haven't fully looked into these though.  

I'm realizing now that Servo may not support all of these as of yet. Is there only specific formats currently?

jdm replied:
For what it's worth, I looked at Firefox's source code and they only ever create tracks with the "main" type as far as I can tell. I don't think we need to try to figure out specifics here until we come across specific use cases where it matters.

sreeise replied:
Yep you are right. I looked at the source code for Firefox before and could never find it then went back and looked after you mentioned it, and there is a 3 year old open bug for multiple audio/video tracks.

Sounds good, I will finish up what can me implemented and send another push. Thanks.

sreeise commented:
Hello, I am sorry for getting back to this so late. I added a part of the media processing steps but I am currently stuck on the kind attribute. I added an inline comment discussing the video and audio formats and how they are determined for this.

Thanks for the help.

jdm commented:
```
error: unused import: `crate::dom::htmlaudioelement::HTMLAudioElement`
  --> components/script/dom/htmlmediaelement.rs:32:5
   |
32 | use crate::dom::htmlaudioelement::HTMLAudioElement;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`
```

jdm commented on components/script/dom/audiotracklist.rs:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.tracks.borrow().is_empty().clone()
```
No need to clone this value.

jdm commented on components/script/dom/audiotracklist.rs:
```diff
@@ -0,0 +1,104 @@
+    event_handler!(onaddtrack, GetOnaddtrack, SetOnaddtrack);
+
+    // https://html.spec.whatwg.org/multipage/#handler-tracklist-onremovetrack
+    event_handler!(onremovetrack, GetOnremovetrack, SetOnremovetrack);
```
Remove the `on` prefix from the first argument for event handlers.

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1233,74 @@ impl HTMLMediaElement {
+                        }
+
+                        // Steps 7
+                        // Fire an event named addtrack at this AudioTrackList object, using TrackEvent
```
Can we implement this step in this PR?

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1233,74 @@ impl HTMLMediaElement {
+                        }
+
+                        // Steps 7
+                        // Fire an event named addtrack at this VideoTrackList object, using TrackEvent
```
Can we implement this step in this PR?

jdm commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,118 @@
+                self.set_selected(idx, false);
+            }
+        }
+        self.tracks.borrow_mut().push(Dom::from_ref(track));
```
We should push this before calling set_selected, otherwise the event handler for the change event could see an unexpected intermediate state.

jdm commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,118 @@
+    event_handler!(onaddtrack, GetOnaddtrack, SetOnaddtrack);
+
+    // https://html.spec.whatwg.org/multipage/#handler-tracklist-onremovetrack
+    event_handler!(onremovetrack, GetOnremovetrack, SetOnremovetrack);
```
Remove the `on` prefix from the first argument to these event handlers.

jdm commented on components/script/dom/audiotracklist.rs:
```diff
@@ -0,0 +1,104 @@
+            .position(|track| track.enabled())
+    }
+
+    pub fn set_enabled(&self, idx: usize, value: bool) {
```
I filed https://github.com/servo/servo/issues/22799 to make this interact with the media backend. Could you add a TODO comment linking to that issue here and in the equivalent video code?

jdm commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,118 @@
+        if let Some(track) = self.item(idx) {
+            track.set_selected(value);
+        }
+        self.upcast::<EventTarget>().fire_event(atom!("change"));
```
Per https://html.spec.whatwg.org/multipage/media.html#toggle-audio-track, this needs to be a queued task like https://github.com/servo/servo/blob/f1dd31f70440fa9c7a40525bd1e03eede568f74d/components/script/dom/htmlmediaelement.rs#L328-L334. Also, we should only queue the task if an actual change occurred.

jdm reviewed (changes requested):
This is very close!

sreeise commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1234,118 @@ impl HTMLMediaElement {
+                            let task_source = window.task_manager().media_element_task_source();
+                            let this = Trusted::new(self);
+
+                            task_source
```
Is this the correct way to implement the task source event? I am not familiar with the naming convention for the tasks or if this is specific in some way.

sreeise replied:
Some other areas will also queue this task and I am wondering whether these should take place in AudioTrackList and VideoTrackList?

sreeise commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,76 @@
+    pub fn Constructor(
+        window: &Window,
+        type_: DOMString,
+        init: &TrackEventBinding::TrackEventInit,
```
The trait here requires an `&Option<UnionType>` for audio/video/text tracks which cant be copied or dereferenced from what I can tell. The GetTrack method, however, does not return a reference so I am assuming here that this needs to be wrapped in some type but I am unsure as to what type that is or if there is a specific way to implement an event interface with UnionTypes?

sreeise commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    pub fn add(&self, track: &VideoTrack) {
+        self.tracks.borrow_mut().push(Dom::from_ref(track));
```
You may have mentioned this before but just to get some clarification, according to the spec it looks like VideoTrackList needs to somehow track the video tracks that were previously selected:

> Whenever a track in a VideoTrackList that was previously not selected is selected, and whenever the selected track in a VideoTrackList is unselected without a new track being selected in its stead, the user agent must queue a task to fire an event named change at the VideoTrackList object. This task must be queued before the task that fires the resize event, if any.

This seems like a field be added, such as a vector to track index's or a boolean field for a video track for being previously selected but I may be overthinking here.

jdm replied:
We can either add a new index member that stores the currently selected index, or we can scan the list of tracks each time.

sreeise commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1280,7 +1406,43 @@ impl HTMLMediaElement {
+                        .queue(
+                            task!(media_track_change: move || {
+                                let this = this.root();
+                                this.AudioTracks().set_selected(0, true);
```
I completely missed this. Should have been `this.AudioTracks().set_enabled(0, true)`

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1234,118 @@ impl HTMLMediaElement {
+                        // Step 5. & 6,
+                        if audio_track.enabled() || self.AudioTracks().enabled_index().is_none() {
+                            let window = window_from_node(self);
+                            let task_source = window.task_manager().media_element_task_source();
```
If you use `media_element_task_source_with_canceller` then we can call `queue_with_canceller` instead of `queue`.

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1234,118 @@ impl HTMLMediaElement {
+                                    task!(media_track_change: move || {
+                                        let this = this.root();
+                                        this.AudioTracks().set_enabled(0, true);
+                                        this.upcast::<EventTarget>().fire_event(atom!("change"));
```
This doesn't quite implement the spec as written - the task is supposed to only fire the event; the actual change is status should have occurred synchronously before queueing the task. Additionally, the change event should be fired at the AudioTrackList, rather than the media element (per https://html.spec.whatwg.org/multipage/media.html#dom-audiotrack-enabled). I think it would make sense to move the event queueing into set_enabled, rather than in this algorithm.

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1234,118 @@ impl HTMLMediaElement {
+                            let this = Trusted::new(self);
+
+                            task_source
+                                .queue(
```
Same comments about this code as the audio one per https://html.spec.whatwg.org/multipage/media.html#dom-videotrack-selected. The event code should be part of set_selected, rather than this algorithm.

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1234,118 @@ impl HTMLMediaElement {
+                        }
+
+                        // Steps 7.
+                        TrackEvent::new(
```
This creates an event object but does not do anything with it. We should store it in a variable and use the [Event::fire](https://doc.servo.org/script/dom/event/struct.Event.html#method.fire) API to dispatch it at the track list.

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1220,6 +1234,118 @@ impl HTMLMediaElement {
+                        }
+
+                        // Steps 7.
+                        TrackEvent::new(
```
Same comment about dispatching the event.

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1280,7 +1406,43 @@ impl HTMLMediaElement {
+                }
+
+                // Step 13
+                if !self.VideoTracks().is_empty() && self.VideoTracks().selected_index().is_none() {
```
Steps 12 and 13 should be calling set_selected/set_enabled only; the DOM event code should be part of that implementation per previous comments.

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,76 @@
+}
+
+impl TrackEvent {
+    pub fn new_inherited(_track: &Option<VideoTrackOrAudioTrackOrTextTrack>) -> TrackEvent {
```
No need for this to be public.

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,76 @@
+#[dom_struct]
+pub struct TrackEvent {
+    event: Event,
+    #[ignore_malloc_size_of = "Defined in rust-mozjs"]
```
This isn't quite true; the comment should be `Defined by union codegen`

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,76 @@
+pub struct TrackEvent {
+    event: Event,
+    #[ignore_malloc_size_of = "Defined in rust-mozjs"]
+    track: Option<VideoTrackOrAudioTrackOrTextTrack>,
```
We should define our own enum:
```rust
enum MediaTrack {
    Video(Dom<VideoTrack>),
    Audio(Dom<AudioTrack>),
    Text(Dom<TextTrack>),
}
```
and use that in this member instead.

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,76 @@
+
+impl TrackEventMethods for TrackEvent {
+    // https://html.spec.whatwg.org/multipage/#dom-trackevent-track
+    fn GetTrack(&self) -> Option<VideoTrackOrAudioTrackOrTextTrack> {
```
We can create a new instance of `VideoTrackOrAudioTrackOrTextTrack` (which is an enum; see [the docs](https://doc.servo.org/script/dom/bindings/codegen/UnionTypes/index.html) for other examples) based on the value stored in `self.track` and calling `DomRoot::from_ref`.

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,76 @@
+
+impl TrackEvent {
+    pub fn new_inherited(_track: &Option<VideoTrackOrAudioTrackOrTextTrack>) -> TrackEvent {
+        unimplemented!();
```
We can match on the input enum and call `Dom::from_ref` to store in a new instance of the enum I suggested adding.

sreeise commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -1280,7 +1381,11 @@ impl HTMLMediaElement {
 
-                // XXX Steps 12 and 13 require audio and video tracks support.
+                // Step 12
+                self.AudioTracks().set_enabled(0, true);
```
On steps 12 and 13 you previously mentioned that they should call only set_enabled/set_selected. I just wanted to get clarification on this. I am probably wrong here but it seems to me if we don't check whether an audio track enabled attribute or video track selected attribute is already enabled/selected respectfully wouldn't this cause a previously selected video track to be deselected if it is not at index 0 or an audio track to be enabled regardless of whether one was enabled previously? The specification says to only enable and audio track or select a video track if there is not an enabled audio or selected video track when arriving at these steps.

jdm replied:
I agree with you. I don't think we actually need this at all, since the code that processes video and audio tracks has already checked if no track has been selected/enabled and done so. I would replace these two lines with `// Step 12 & 13 are already handled by the earlier media track processing.`

sreeise commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,115 @@
+use servo_atoms::Atom;
+
+#[must_root]
+#[derive(JSTraceable, MallocSizeOf)]
```
I am not very familiar with how/where to define the MallocSizeOf trait or ignored attribute as well as must_root/unrooted_must_root. I assumed here that since a new enum is created it should define MallocSizeOf or should the MediaTrack in TrackEvent still have the ignore_malloc_of_size above it?

jdm replied:
Deriving as often as possible is the best choice.

sreeise commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    // TODO(#22799) Integrate DOM Audio and Video track selection with media player
+    pub fn set_selected(&self, idx: usize, value: bool) {
```
Hopefully I got the logic correct here on selecting the video track. I am thinking it can be cleaned up more though.

sreeise commented:
I am getting a build warning for rustc_on_umplimented:

>  --> components/script/dom/bindings/conversions.rs:74:5
>    |
> 74 |     rustc_on_unimplemented = "The IDL interface `{Self}` is not derived from `{T}`."
>    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = note: #[warn(unused_attributes)] on by default

I assume this has something to do with how TrackEvent is implemented. Do you have any info on what I should do to fix this?

jdm commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,172 @@
+                    );
+                }
+            }
+        }
```
I recommend restructuring this code as follows:
```rust
// get chosen track or return early
// if the chosen track's selection status matches the new status, return early
// get the currently selected track and deselect it
// if we are selecting a track, select the chosen track
// queue a change event
```
Does that make sense?

sreeise replied:
Yes thank you. The specification is somewhat confusing on this.

sreeise commented on components/script/dom/videotrack.rs:
```diff
@@ -0,0 +1,97 @@
+
+    // https://html.spec.whatwg.org/multipage/#dom-videotrack-selected
+    fn SetSelected(&self, value: bool) {
+        self.set_selected(value);
```
That works for me.

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,137 @@
+        ))
+    }
+
+    pub fn queue_track_event(&self) {
```
Given that this is only called by one location, I don't think we need a method for this operation.

jdm commented on components/script/dom/trackevent.rs:
```diff
@@ -0,0 +1,137 @@
+            task!(track_event_queue: move || {
+                let this = this.root();
+                let global = this.global();
+                this.upcast::<Event>().fire(global.upcast::<EventTarget>());
```
Not quite; per https://html.spec.whatwg.org/multipage/media.html#dom-media-addtexttrack we need to fire the event at the TextTrackList object, not the global.

jdm commented on components/script/dom/videotracklist.rs:
```diff
@@ -0,0 +1,146 @@
+            if current != idx {
+                // Set the tracks selected status.
+                self.tracks.borrow()[current].set_selected(false);
+                track.set_selected(true);
```
This code ignores the case where we are deselecting the currently selected track. It also ignores the case where we have no currently selected track and we select one.

jdm reviewed (changes requested):
Much closer!

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit f2fcc88 has been approved by `jdm`

<!-- @bors-servo r=jdm f2fcc887722a1738c20cac3cb39374cece608a03 -->

bors-servo commented:
:hourglass: Testing commit f2fcc887722a1738c20cac3cb39374cece608a03 with merge e0065c0b4eb787a81123af37d88aeb7801b91fa4...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11678)

CYBAI commented:
I think test expectation for the `html/semantics/embedded-content/media-elements/interfaces/TrackEvent/constructor.html.ini` needs to be updated.
Besides, I think the test of last `FAIL` one should be updated.

```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "TrackEvent constructor, one arg", 
    "test": "/html/semantics/embedded-content/media-elements/interfaces/TrackEvent/constructor.html", 
    "line": 123663, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "TrackEvent constructor, two args", 
    "test": "/html/semantics/embedded-content/media-elements/interfaces/TrackEvent/constructor.html", 
    "line": 123664, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "TrackEvent created with createEvent", 
    "test": "/html/semantics/embedded-content/media-elements/interfaces/TrackEvent/createEvent.html", 
    "line": 201645, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_true: If this is failing: DANGER, are you sure you want to expose the new interface AudioTrack to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false", 
    "stack": "test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:70:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1587:20\ntest@http://web-platform.test:8000/resources/testharness.js:544:21\ntest_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3\n@http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1\n", 
    "subtest": "Interfaces exposed on the window", 
    "test": "/_mozilla/mozilla/interfaces.html", 
    "line": 212880, 
    "action": "test_result", 
    "expected": "PASS"
}
```

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4c79042 has been approved by `jdm`

<!-- @bors-servo r=jdm 4c7904235db36ae38d64d3bdfe7549b4c52344c7 -->

bors-servo commented:
:hourglass: Testing commit 4c7904235db36ae38d64d3bdfe7549b4c52344c7 with merge 2f92916f4a9e06665571b1563e33aa8a1ad6848c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11741)

jdm commented:
Ok, this should be the last round of test expectation updates!

```
  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: document.createElement("video") must inherit property "audioTracks" with the proper type

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: document.createElement("video") must inherit property "videoTracks" with the proper type

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: document.createElement("audio") must inherit property "audioTracks" with the proper type

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: document.createElement("audio") must inherit property "videoTracks" with the proper type

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: new Audio() must inherit property "audioTracks" with the proper type

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: new Audio() must inherit property "videoTracks" with the proper type

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: attribute audioTracks

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLMediaElement interface: attribute videoTracks

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: existence and properties of interface object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface object name

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: existence and properties of interface prototype object's "constructor" property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: existence and properties of interface prototype object's @@unscopables property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: attribute length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: operation getTrackById(DOMString)

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: attribute onchange

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: attribute onaddtrack

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrackList interface: attribute onremovetrack

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: existence and properties of interface object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface object name

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: existence and properties of interface prototype object's "constructor" property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: existence and properties of interface prototype object's @@unscopables property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: attribute id

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: attribute kind

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: attribute label

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: attribute language

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] AudioTrack interface: attribute enabled

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: existence and properties of interface object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface object name

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: existence and properties of interface prototype object's "constructor" property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: existence and properties of interface prototype object's @@unscopables property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: attribute length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: operation getTrackById(DOMString)

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: attribute selectedIndex

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: attribute onchange

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: attribute onaddtrack

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrackList interface: attribute onremovetrack

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: existence and properties of interface object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface object name

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: existence and properties of interface prototype object's "constructor" property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: existence and properties of interface prototype object's @@unscopables property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: attribute id

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: attribute kind

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: attribute label

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: attribute language

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] VideoTrack interface: attribute selected

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface: existence and properties of interface object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface object length

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface object name

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface: existence and properties of interface prototype object's "constructor" property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface: existence and properties of interface prototype object's @@unscopables property

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface: attribute track

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent must be primary interface of new TrackEvent("addtrack", {track:document.createElement("track").track})

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] Stringification of new TrackEvent("addtrack", {track:document.createElement("track").track})

  ▶ Unexpected subtest result in /html/dom/interfaces.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] TrackEvent interface: new TrackEvent("addtrack", {track:document.createElement("track").track}) must inherit property "track" with the proper type
```

sreeise commented:
> Ok, this should be the last round of test expectation updates!

No worries, I was thinking these might fail : ).

sreeise commented:
Hopefully the changes are correct. I am having an issue testing these locally. When I try run `tests/wpt/web-platform-tests/html/dom/interfaces.https.html` it crashes with this error:

```
pid:7447 VMware, Inc.
 0:03.95 pid:7447 softpipe
 0:03.95 pid:7447 3.3 (Core Profile) Mesa 18.3.0-devel
 0:03.96 pid:7447 assertion failed: `(left == right)`
 0:03.96 pid:7447   left: `1282`,
 0:03.96 pid:7447  right: `0`
```
which begins at `.cargo/registry/src/github.com-1ecc6299db9ec823/gleam-0.6.8/src/gl.rs:93`

The other thing I noticed on development builds, which happened before I changed the test expectations, is this warning:

```
 --> components/script/dom/bindings/conversions.rs:74:5
   |
74 |     rustc_on_unimplemented = "The IDL interface `{Self}` is not derived from `{T}`."
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_attributes)] on by default
```
I am not sure what the cause of these are. They could just be happening for me locally.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 094777c has been approved by `jdm`

<!-- @bors-servo r=jdm 094777c77cc4f9b4ffc6057224153a36c47bd8db -->

bors-servo commented:
:hourglass: Testing commit 094777c77cc4f9b4ffc6057224153a36c47bd8db with merge 543bf197298b6942dd0dba576f3bf17a4356c21f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11747)

sreeise commented:
Yep I missed one, sorry!. Hopefully this one will finally pass.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit cac4aa5 has been approved by `jdm`

<!-- @bors-servo r=jdm cac4aa56f77c46db75c9ca1bfdf23be39cfd4604 -->

bors-servo commented:
:hourglass: Testing commit cac4aa56f77c46db75c9ca1bfdf23be39cfd4604 with merge 4ccbd1e87eb81d6cdb1658aadf33cc0795be5adf...

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/1155), [arm32](http://build.servo.org/builders/arm32/builds/14031), [arm64](http://build.servo.org/builders/arm64/builds/14059), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11694), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11749), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6636), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6605), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10847), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10910), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6660), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6593), [magicleap](http://build.servo.org/builders/magicleap/builds/774), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/TIk-3nzkQ2iR0NdqZbq9PQ)
Approved by: jdm
Pushing 4ccbd1e87eb81d6cdb1658aadf33cc0795be5adf to master...

jdm commented:
Woohoo! Thanks for sticking with this!
